compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,505 @@
+  SmallVector<int64_t> innerPerm =
+      getPackUnpackNormalizedPerm(unpackedRank, innerDimsPos);
+  applyPermutationToVector<int64_t>(innerDims, innerPerm);
+
+  // Ditto for the outer dims.
+  SmallVector<int64_t> perm = outerDims;
+
+  rankReducedOuterDimsPerm =
+      getPackUnpackNormalizedPerm(unpackedRank, rankReducedOuterDimsPerm);
+  if (!rankReducedOuterDimsPerm.empty())
+    applyPermutationToVector<int64_t>(perm, rankReducedOuterDimsPerm);
+
+  // The tile always ends up as the inner most dims after packing.
+  perm.append(innerDims);
+
+  return perm;
+}
+
+class GeneralizePackOpPattern
+    : public OpRewritePattern<IREE::LinalgExt::PackOp> {
+ public:
+  using OpRewritePattern<IREE::LinalgExt::PackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::LinalgExt::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    if (llvm::any_of(packOp.getMixedTiles(),
+                     [](OpFoldResult tile) { return tile.is<Value>(); })) {
+      return rewriter.notifyMatchFailure(
+          packOp, "require inner tile sizes being static");
+    }
+
+    auto innerDimsPos = packOp.getInnerDimsPos();
+    int64_t srcRank = packOp.getInputRank();
+    auto destShape = packOp.getOutputType().getShape();
+    if (llvm::any_of(innerDimsPos, [destShape](int64_t index) {
+          return destShape[index] != 1;
+        })) {
+      return rewriter.notifyMatchFailure(
+          packOp,
+          "require the tiled outer dimensions of the result are all 1s");
+    }
+
+    // 1. Use rank-reduced memref.subview op to extract the tile and untiled
+    // outer dims.
+    Location loc = packOp.getLoc();
+    Value input = getPackOpInputOrPaddedSource(rewriter, packOp);
+    auto inputShape = packOp.getInputType().getShape();
+    DenseMap<int64_t, OpFoldResult> dimAndTileMapping =
+        packOp.getDimAndTileMapping();
+    Attribute zeroIdxAttr = rewriter.getIndexAttr(0);
+    Attribute oneIdxAttr = rewriter.getIndexAttr(1);
+    SmallVector<OpFoldResult> readOffsets(srcRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> readStrides(srcRank, oneIdxAttr);
+    SmallVector<OpFoldResult> readSizes;
+    SmallVector<int64_t> readShape;
+    for (auto i : llvm::seq<unsigned>(0, srcRank)) {
+      if (dimAndTileMapping.count(i)) {
+        readShape.push_back(getConstantIntValue(dimAndTileMapping[i])
+                                .value_or(ShapedType::kDynamic));
+        readSizes.push_back(dimAndTileMapping[i]);
+        continue;
+      }
+      if (ShapedType::isDynamic(inputShape[i])) {
+        assert(false);
+        // TODO: Dynamic input shape
+      } else {
+        readSizes.push_back(rewriter.getIndexAttr(inputShape[i]));
+      }
+      readShape.push_back(inputShape[i]);
+    }
+
+    Type elemType = packOp.getInputType().getElementType();
+    Attribute memorySpace =
+        packOp.getInputType().cast<MemRefType>().getMemorySpace();
+    auto readType = MemRefType::get(
```

Comment:
I am not even sure why this subview is needed here? This is essentially taking the full slice of the underlying object?

Reply:
Agreed. Removed.